EU4toV2/Source/V2World/V2World.cpp

```diff
@@ -1287,51 +1287,49 @@ std::optional<std::string> V2::World::determineProvinceControllership(const std:
 
 void V2::World::setupStates()
 {
-	std::list<std::shared_ptr<Province>> unassignedProvs;
+	/* std::list<std::shared_ptr<Province>> unassignedProvs;
 	for (const auto& province: provinces)
-		unassignedProvs.push_back(province.second);
+		unassignedProvs.push_back(province.second);*/
 
-	while (!unassignedProvs.empty())
-	{
-		auto iter = unassignedProvs.begin();
-		const auto provId = (*iter)->getID();
-		auto owner = (*iter)->getOwner();
+	std::set<int> provinceInState;
 
-		if (owner.empty())
+	for (const auto& mapEntry: provinces)
+	{
+		auto province = mapEntry.second;
+		const auto provId = province->getID();
+		auto owner = province->getOwner();
+		if (owner.empty() || provinceInState.contains(provId))
 		{
-			unassignedProvs.erase(iter);
 			continue;
 		}
 
-		auto newState = std::make_shared<State>(stateId, *iter);
-		stateId++;
-		auto neighbors = stateMapper.getAllProvincesInState(provId);
-
 		// We are breaking states apart according to colonial status. This is so primitives can retain
 		// their full states next to colonizers who have colonial provinces in the same state.
 		// This ALSO means multiple naval bases within apparently single state.
-		const auto colonial = (*iter)->isColony();
-		newState->setColonial(colonial);
-		iter = unassignedProvs.erase(iter);
+		auto isColony = province->isColony();
+		auto newState = std::make_shared<State>(stateId, province);
+		newState->setColonial(isColony);
+		stateId++;
+
+		provinceInState.insert(provId);
 
-		for (const auto& neighborID: neighbors)
+		auto neighborIds = stateMapper.getAllProvincesInState(provId);
+		for (const auto& neighborId: neighborIds)
 		{
-			for (iter = unassignedProvs.begin(); iter != unassignedProvs.end(); ++iter)
+
+			auto neighbor = provinces[neighborId];
+			if (neighborId != provId)
 			{
-				if ((*iter)->getID() == neighborID)
+				if (owner == neighbor->getOwner())
 				{
-					if ((*iter)->getOwner() == owner)
+					if (isColony == neighbor->isColony())
```

Comment:
you can merge these 2 ifs into a single one, and you're also lacking the critical third one: if (!provinceInState.contains(neighborID)).

This is relevant on m-to-n mappings, you can easily add the neighbor to multiple states or multiple times to a single state this way.